ykrt/src/compile/jitc_yk/jit_ir.rs

```diff
@@ -28,6 +53,75 @@ const OPERAND_IDX_MASK: u16 = 0x7fff;
 // The largest operand index we can express in 15 bits.
 const MAX_OPERAND_IDX: u16 = (1 << 15) - 1;
 
+/// [Instruction] to [InstrIdx] mapping used for stringifying instructions.
+///
+/// See [JitIRDisplay] for why this is required.
+pub(crate) struct LocalNumbers<'a> {
+    m: &'a Module,
+}
+
+impl<'a> LocalNumbers<'a> {
+    fn new(m: &'a Module) -> Result<LocalNumbers<'a>, CompilationError> {
+        Ok(Self { m })
+    }
+
+    /// Returns the [InstrIdx] of the specified [Instruction] in the [Module].
+    ///
+    /// # Panics
+    ///
+    /// Panics if the instruction isn't present in the module.
+    fn idx(&self, instr: &Instruction) -> Result<InstrIdx, CompilationError> {
+        // FIXME: This is inefficient.
+        for (idx, candidate) in self.m.instrs().iter().enumerate() {
+            if ptr::addr_eq(instr, candidate) {
+                return Ok(InstrIdx::new(idx)?);
+            }
+        }
+        panic!(); // Not found!
+    }
+}
+
+/// A trait for stringifying JIT IR.
+///
+/// We use this instead of [std::fmt::Display] because many IR constructs can't print something
+/// human-readable without querying the [Module] for more information. For example for to convert
+/// type indices into a human-readable type description. Hence we need the ability pass down a
+/// reference to a module.
+///
+/// Stringification also requires a [LocalNumbers] mapping. This is because when you want to print
+/// an [Instruction], we want to print something like `%4: i32 = Add %1, %2`. But an instruction on
+/// its own doesn't know its instruction index in order to print `%4`. [LocalNumbers] lets an
+/// instruction find its index in the module.
+pub(crate) trait JitIRDisplay {
+    /// Return a human-readable string.
+    ///
+    /// This is the outward-facing API for stringification. Don't call this from other
+    /// `JitIRDisplay` implementations. Use [JitIRDisplay::to_string_impl()] instead and pass down
+    /// the existing output string and [LocalNumbers] struct.
+    fn to_string(&self, m: &Module) -> Result<String, CompilationError> {
+        let mut s = String::new();
+        self.to_string_impl(m, &mut s, &LocalNumbers::new(m)?)?;
+        Ok(s)
+    }
+
+    /// This method does the actual stringification of an IR element.
+    fn to_string_impl<'a>(
+        &self,
+        m: &Module,
+        s: &mut String,
+        nums: &LocalNumbers<'a>,
+    ) -> Result<(), CompilationError>;
+
+    /// Print myself to stderr in human-readable form.
```

Comment:
"Print self to..."?

Reply:
fixed in aa044fa7024bc9831c8a804719d2ef054c41d70a